core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+            assert_eq!(
+                write_thread_id, &thread_id,
+                "outstanding write lock must be on same thread"
+            );
+        }
+    }
+
+    /// Unlocks the given `account` for reading on `thread_id`.
+    /// Panics if the account is not locked for reading on `thread_id`.
+    fn read_unlock_account(&mut self, account: &Pubkey, thread_id: ThreadId) {
+        match self.read_locks.entry(*account) {
+            Entry::Occupied(mut entry) => {
+                let (thread_set, lock_counts) = entry.get_mut();
+                assert!(
+                    thread_set.contains(thread_id),
+                    "outstanding read lock must be on same thread"
+                );
+                lock_counts[thread_id] -= 1;
+                if lock_counts[thread_id] == 0 {
+                    thread_set.remove(thread_id);
+                    if thread_set.is_empty() {
+                        entry.remove();
+                    }
+                }
+            }
+            Entry::Vacant(_) => {
+                panic!("read lock must exist for account: {account}");
+            }
+        }
+    }
+}
+
+impl BitAndAssign for ThreadSet {
+    fn bitand_assign(&mut self, rhs: Self) {
+        self.0 &= rhs.0;
+    }
+}
+
+impl Sub for ThreadSet {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(self.0 & !rhs.0)
+    }
+}
+
+#[cfg(test)]
+static_assertions::assert_eq_size!(ThreadSet, u64); // Formatting must change if size changes
+impl Display for ThreadSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ThreadSet({:#064b})", self.0)
+    }
+}
+
+impl Debug for ThreadSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+impl ThreadSet {
+    #[inline(always)]
+    pub fn none() -> Self {
+        Self(0)
+    }
+
+    #[inline(always)]
+    pub fn any(num_threads: usize) -> Self {
+        Self((1 << num_threads) - 1)
+    }
+
+    #[inline(always)]
+    pub fn only(thread_id: ThreadId) -> Self {
+        Self(1 << thread_id)
+    }
+
+    #[inline(always)]
+    pub fn num_threads(&self) -> u8 {
+        self.0.count_ones() as u8
+    }
+
+    #[inline(always)]
+    pub fn only_one_scheduled(&self) -> Option<ThreadId> {
+        (self.num_threads() == 1).then_some(self.0.trailing_zeros() as ThreadId)
+    }
+
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+
+    #[inline(always)]
+    pub fn contains(&self, thread_id: ThreadId) -> bool {
+        self.0 & (1 << thread_id) != 0
+    }
+
+    #[inline(always)]
+    pub fn insert(&mut self, thread_id: ThreadId) {
+        self.0 |= 1 << thread_id;
+    }
+
+    #[inline(always)]
+    pub fn remove(&mut self, thread_id: ThreadId) {
+        self.0 &= !(1 << thread_id);
+    }
+
+    #[inline(always)]
+    pub fn threads_iter(self) -> impl Iterator<Item = ThreadId> {
```

Comment:
nit: naming: `contained_threads()`?

Reply:
Maybe misguided but I like to be explicit about the fact it's returning an iterator.
I changed it to `contained_threads_iter()`: 90a77499438964130186b312b4c34758e559bd6b